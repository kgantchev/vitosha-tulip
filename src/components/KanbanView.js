// src/components/KanbanView.js
import React from 'react';
import dayjs from 'dayjs';
import {
    Box,
    Typography,
    Card,
    CardContent,
    CardMedia,
} from '@mui/material';

const KanbanView = ({ kanbanColumns = {} }) => {
    const columnOrder = ['to do', 'definition', 'in progress', 'in review', 'complete'];

    return (
        <Box>
            {Object.keys(kanbanColumns).map((listName) => (
                <Box key={listName} mb={4}>
                    <Typography variant="h4" gutterBottom>
                        {listName}
                    </Typography>
                    <Box
                        sx={{
                            display: 'flex',
                            flexWrap: 'wrap', // Allow columns to wrap to the next line
                            gap: 2,
                        }}
                    >
                        {columnOrder.map((columnName) => {
                            const tasks = kanbanColumns[listName][columnName] || [];
                            return (
                                <Box
                                    key={columnName}
                                    sx={{
                                        flex: '1 1 200px',
                                        minWidth: '200px',
                                    }}
                                >
                                    <Typography
                                        variant="h6"
                                        gutterBottom
                                        component="h6"
                                        data-testid="column-heading"
                                    >
                                        {columnName.toUpperCase()}
                                    </Typography>
                                    {tasks.map((task) => (
                                        <Card key={task.id} sx={{ mb: 2 }}>
                                            <CardContent>
                                                <Typography variant="subtitle1" fontWeight="bold">
                                                    {task.name}
                                                </Typography>
                                                {task.attachments?.[0] && (
                                                    <CardMedia
                                                        component="img"
                                                        src={task.attachments[0].base64Image}
                                                        alt={task.attachments[0].fileName}
                                                        sx={{ mt: 1, borderRadius: 1 }}
                                                    />
                                                )}
                                                <Typography variant="caption" color="textSecondary">
                                                    Last updated: {dayjs(Number(task.date_updated)).format('YYYY-MM-DD')}
                                                </Typography>
                                            </CardContent>
                                        </Card>
                                    ))}
                                </Box>
                            );
                        })}
                    </Box>
                </Box>
            ))}
        </Box>
    );
};

export default KanbanView;