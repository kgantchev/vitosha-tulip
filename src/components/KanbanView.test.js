// src/components/KanbanView.test.js
import React from 'react';
import { render, screen, within } from '@testing-library/react';
import KanbanView from './KanbanView';
import '@testing-library/jest-dom';
import dayjs from 'dayjs';

// Mock Data for Kanban Columns
const mockKanbanColumns = {
    "Project Alpha": {
        "to do": [
            {
                id: '1',
                name: 'Design the landing page',
                date_updated: String(dayjs('2023-10-01').valueOf()),
                attachments: [
                    {
                        base64Image: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...',
                        fileName: 'design.png',
                    },
                ],
            },
            {
                id: '2',
                name: 'Set up CI/CD pipeline',
                date_updated: String(dayjs('2023-10-03').valueOf()),
            },
        ],
        "in progress": [
            {
                id: '3',
                name: 'Develop authentication module',
                date_updated: String(dayjs('2023-10-05').valueOf()),
                attachments: [
                    {
                        base64Image: 'data:image/png;base64,abcd1234...',
                        fileName: 'auth_flow.png',
                    },
                ],
            },
        ],
    },
    "Project Beta": {
        "to do": [],
        "complete": [
            {
                id: '4',
                name: 'Market research',
                date_updated: String(dayjs('2023-09-20').valueOf()),
            },
        ],
    },
};

describe('KanbanView Component', () => {
    test('renders without crashing when kanbanColumns is empty', () => {
        render(<KanbanView kanbanColumns={{}} />);

        // Since kanbanColumns is empty, no list names should be rendered
        const listNameElements = screen.queryAllByRole('heading', { level: 4 });
        expect(listNameElements.length).toBe(0);
    });

    test('renders list names correctly', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // Check for "Project Alpha" and "Project Beta" headings
        const projectAlpha = screen.getByRole('heading', { name: 'Project Alpha' });
        const projectBeta = screen.getByRole('heading', { name: 'Project Beta' });

        expect(projectAlpha).toBeInTheDocument();
        expect(projectBeta).toBeInTheDocument();
    });

    test('renders columns in the specified order', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        const expectedOrder = ['TO DO', 'DEFINITION', 'IN PROGRESS', 'IN REVIEW', 'COMPLETE'];

        Object.keys(mockKanbanColumns).forEach((listName) => {
            const listHeading = screen.getByRole('heading', { name: listName });
            const listSection = listHeading.parentElement;
            const listWithin = within(listSection);

            const renderedColumnNames = [];

            expectedOrder.forEach((columnName) => {
                const columnHeading = listWithin.getByText(columnName, { selector: 'h6' });
                expect(columnHeading).toBeInTheDocument();
                renderedColumnNames.push(columnHeading.textContent.trim());
            });

            expect(renderedColumnNames).toEqual(expectedOrder);
        });
    });

    test('renders tasks within the correct columns', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // Check for tasks in "Project Alpha" -> "TO DO"
        const projectAlphaSection = screen.getByRole('heading', { name: 'Project Alpha' }).parentElement;
        const alphaWithin = within(projectAlphaSection);

        expect(alphaWithin.getByText('TO DO')).toBeInTheDocument();
        expect(alphaWithin.getByText('Design the landing page')).toBeInTheDocument();
        expect(alphaWithin.getByText('Set up CI/CD pipeline')).toBeInTheDocument();

        // Check for tasks in "Project Alpha" -> "IN PROGRESS"
        expect(alphaWithin.getByText('IN PROGRESS')).toBeInTheDocument();
        expect(alphaWithin.getByText('Develop authentication module')).toBeInTheDocument();

        // Check for tasks in "Project Beta" -> "COMPLETE"
        const projectBetaSection = screen.getByRole('heading', { name: 'Project Beta' }).parentElement;
        const betaWithin = within(projectBetaSection);

        expect(betaWithin.getByText('COMPLETE')).toBeInTheDocument();
        expect(betaWithin.getByText('Market research')).toBeInTheDocument();
    });

    test('renders attachments when present', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // Task with attachment: "Design the landing page"
        const designTaskImage = screen.getByAltText('design.png');
        expect(designTaskImage).toBeInTheDocument();
        expect(designTaskImage).toHaveAttribute('src', 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA...');

        // Task with attachment: "Develop authentication module"
        const authTaskImage = screen.getByAltText('auth_flow.png');
        expect(authTaskImage).toBeInTheDocument();
        expect(authTaskImage).toHaveAttribute('src', 'data:image/png;base64,abcd1234...');
    });

    test('does not render attachments when not present', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // Task without attachment: "Set up CI/CD pipeline"
        const ciTaskImage = screen.queryByAltText('Set up CI/CD pipeline');
        expect(ciTaskImage).not.toBeInTheDocument();

        // Task without attachment: "Market research"
        const marketTaskImage = screen.queryByAltText('Market research');
        expect(marketTaskImage).not.toBeInTheDocument();
    });

    test('formats and displays the last updated date correctly', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // "Design the landing page" updated on 2023-10-01
        const designTaskDate = screen.getByText('Last updated: 2023-10-01');
        expect(designTaskDate).toBeInTheDocument();

        // "Set up CI/CD pipeline" updated on 2023-10-03
        const ciTaskDate = screen.getByText('Last updated: 2023-10-03');
        expect(ciTaskDate).toBeInTheDocument();

        // "Develop authentication module" updated on 2023-10-05
        const authTaskDate = screen.getByText('Last updated: 2023-10-05');
        expect(authTaskDate).toBeInTheDocument();

        // "Market research" updated on 2023-09-20
        const marketTaskDate = screen.getByText('Last updated: 2023-09-20');
        expect(marketTaskDate).toBeInTheDocument();
    });

    test('handles multiple list names correctly', () => {
        render(<KanbanView kanbanColumns={mockKanbanColumns} />);

        // Ensure both "Project Alpha" and "Project Beta" sections are present
        const projectAlpha = screen.getByRole('heading', { name: 'Project Alpha' });
        const projectBeta = screen.getByRole('heading', { name: 'Project Beta' });

        expect(projectAlpha).toBeInTheDocument();
        expect(projectBeta).toBeInTheDocument();
    });

    test('matches the snapshot', () => {
        const { asFragment } = render(<KanbanView kanbanColumns={mockKanbanColumns} />);
        expect(asFragment()).toMatchSnapshot();
    });
});