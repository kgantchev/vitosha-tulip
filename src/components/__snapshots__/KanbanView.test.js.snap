// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KanbanView Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1yjvs5a"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
      >
        Project Alpha
      </h4>
      <div
        class="MuiBox-root css-1dqpc7h"
      >
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            TO DO
          </h6>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ypifsw-MuiPaper-root-MuiCard-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-or3kzz-MuiTypography-root"
              >
                Design the landing page
              </h6>
              <img
                alt="design.png"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1h81d90-MuiCardMedia-root"
                src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAA..."
              />
              <span
                class="MuiTypography-root MuiTypography-caption css-1jkzbat-MuiTypography-root"
              >
                Last updated: 2023-10-01
              </span>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ypifsw-MuiPaper-root-MuiCard-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-or3kzz-MuiTypography-root"
              >
                Set up CI/CD pipeline
              </h6>
              <span
                class="MuiTypography-root MuiTypography-caption css-1jkzbat-MuiTypography-root"
              >
                Last updated: 2023-10-03
              </span>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            DEFINITION
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            IN PROGRESS
          </h6>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ypifsw-MuiPaper-root-MuiCard-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-or3kzz-MuiTypography-root"
              >
                Develop authentication module
              </h6>
              <img
                alt="auth_flow.png"
                class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1h81d90-MuiCardMedia-root"
                src="data:image/png;base64,abcd1234..."
              />
              <span
                class="MuiTypography-root MuiTypography-caption css-1jkzbat-MuiTypography-root"
              >
                Last updated: 2023-10-05
              </span>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            IN REVIEW
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            COMPLETE
          </h6>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1yjvs5a"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
      >
        Project Beta
      </h4>
      <div
        class="MuiBox-root css-1dqpc7h"
      >
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            TO DO
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            DEFINITION
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            IN PROGRESS
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            IN REVIEW
          </h6>
        </div>
        <div
          class="MuiBox-root css-1wz9wm8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            data-testid="column-heading"
          >
            COMPLETE
          </h6>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ypifsw-MuiPaper-root-MuiCard-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-or3kzz-MuiTypography-root"
              >
                Market research
              </h6>
              <span
                class="MuiTypography-root MuiTypography-caption css-1jkzbat-MuiTypography-root"
              >
                Last updated: 2023-09-20
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
